Sources/NIOExtras/WritePCAPHandler.swift

```diff
@@ -808,6 +808,131 @@ extension NIOWritePCAPHandler {
+            }
+            let fileHandle: NIOFileHandle = NIOFileHandle(_deprecatedTakingOwnershipOfDescriptor: fd)
+            return AsynchronizedFileSink(fileHandle: fileHandle, eventLoop: eventLoop, errorHandler: errorHandler)
+        }
+        
+        private init(
+            fileHandle: NIOFileHandle,
+            eventLoop: EventLoop,
+            errorHandler: @escaping (Swift.Error) -> Void
+        ) {
+            self.fileHandle = fileHandle
+            self.eventLoop = eventLoop
+            self.errorHandler = errorHandler
+        }
+        
+        public func write(buffer: ByteBuffer) async throws {
+            try await self.eventLoop.submit {
+                try self.fileHandle.withUnsafeFileDescriptor { fd in
+                    var buffer = buffer
+                    while buffer.readableBytes > 0 {
+                        try buffer.readWithUnsafeReadableBytes { dataPtr in
+                            let written = sysWrite(fd, dataPtr.baseAddress, dataPtr.count)
+                            guard written > 0 else {
+                                throw Error(errorCode: Error.ErrorCode.cannotWriteToFileError.rawValue)
+                            }
+                            return written
+                        }
+                    }
+                }
+            }.get()
+        }
+        
+        /// Asynchronously syncs the file to disk using fsync.
+        public func asyncSync() async throws {
+            try await withCheckedThrowingContinuation { continuation in
+                _ = self.eventLoop.submit {
+                    do {
+                        try self.fileHandle.withUnsafeFileDescriptor { fd in
+                            let result: CInt = fsync(fd)
+                            if result != 0 {
+                                throw Error(errorCode: Error.ErrorCode.cannotWriteToFileError.rawValue)
+                            }
+                        }
+                        continuation.resume(returning: ())
+                    } catch {
+                        continuation.resume(throwing: error)
+                    }
+                }
+            }
+        }
+
+        /// Asynchronously closes the file sink.
+        public func close() async throws {
+            try await withCheckedThrowingContinuation { continuation in
+                _ = self.eventLoop.submit {
+                    do {
+                        try self.fileHandle.close()
+                        continuation.resume(returning: ())
+                        print("File successfully closed.")
+                    } catch {
+                        continuation.resume(throwing: error)
+                        print("Error closing file: \(error)")
+                    }
+                }
+            }
+        }
+    }
 }
 
 extension NIOWritePCAPHandler.SynchronizedFileSink: @unchecked Sendable {}
+extension NIOWritePCAPHandler.AsynchronizedFileSink: @unchecked Sendable {}
```

Comment:
Let's put the state in a `NIOLockedValueBox`. Then we can make this checked Sendable.

Reply:
Got it!